FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp

```diff
@@ -901,4 +950,402 @@ void OpDispatchBuilder::AVX128_VPSIGN(OpcodeArgs) {
+}
+
+template<size_t ElementSize>
+void OpDispatchBuilder::AVX128_VPSRA(OpcodeArgs) {
+  AVX128_VectorShiftWideImpl(Op, ElementSize, IROps::OP_VSSHRSWIDE);
+}
+
+template<size_t ElementSize>
+void OpDispatchBuilder::AVX128_VPSLL(OpcodeArgs) {
+  AVX128_VectorShiftWideImpl(Op, ElementSize, IROps::OP_VUSHLSWIDE);
+}
+
+template<size_t ElementSize>
+void OpDispatchBuilder::AVX128_VPSRL(OpcodeArgs) {
+  AVX128_VectorShiftWideImpl(Op, ElementSize, IROps::OP_VUSHRSWIDE);
+}
+
+void OpDispatchBuilder::AVX128_VariableShiftImpl(OpcodeArgs, IROps IROp) {
+  AVX128_VectorBinaryImpl(Op, GetDstSize(Op), GetSrcSize(Op), [this, IROp](size_t ElementSize, Ref Src1, Ref Src2) {
+    DeriveOp(Shift, IROp, _VUShr(OpSize::i128Bit, ElementSize, Src1, Src2, true));
+    return Shift;
+  });
+}
+
+void OpDispatchBuilder::AVX128_VPSLLV(OpcodeArgs) {
+  AVX128_VariableShiftImpl(Op, IROps::OP_VUSHL);
+}
+
+void OpDispatchBuilder::AVX128_VPSRAVD(OpcodeArgs) {
+  AVX128_VariableShiftImpl(Op, IROps::OP_VSSHR);
+}
+
+void OpDispatchBuilder::AVX128_VPSRLV(OpcodeArgs) {
+  AVX128_VariableShiftImpl(Op, IROps::OP_VUSHR);
+}
+
+template<size_t ElementSize>
+void OpDispatchBuilder::AVX128_VPSRLI(OpcodeArgs) {
+  AVX128_VectorShiftImmImpl(Op, ElementSize, IROps::OP_VUSHRI);
+}
+
+template<size_t ElementSize>
+void OpDispatchBuilder::AVX128_VPSLLI(OpcodeArgs) {
+  AVX128_VectorShiftImmImpl(Op, ElementSize, IROps::OP_VSHLI);
+}
+
+template<size_t ElementSize>
+void OpDispatchBuilder::AVX128_VPSRAI(OpcodeArgs) {
+  AVX128_VectorShiftImmImpl(Op, ElementSize, IROps::OP_VSSHRI);
+}
+
+void OpDispatchBuilder::AVX128_ShiftDoubleImm(OpcodeArgs, bool Right) {
+  const auto DstSize = GetDstSize(Op);
+  const auto Is128Bit = DstSize == Core::CPUState::XMM_SSE_REG_SIZE;
+  const uint64_t Shift = Op->Src[1].Literal();
+  const uint64_t ExtrShift = Right ? Shift : OpSize::i128Bit - Shift;
+
+  auto Src = AVX128_LoadSource_WithOpSize(Op, Op->Src[0], Op->Flags, !Is128Bit);
+
+  RefPair Result {};
+  if (Shift == 0) [[unlikely]] {
+    Result = Src;
+  } else if (Shift >= Core::CPUState::XMM_SSE_REG_SIZE) {
+    Result.Low = LoadZeroVector(OpSize::i128Bit);
+    Result.High = Result.High;
+  } else {
+    Ref ZeroVector = LoadZeroVector(OpSize::i128Bit);
+    RefPair Zero {ZeroVector, ZeroVector};
+    RefPair Src1 = Right ? Zero : Src;
+    RefPair Src2 = Right ? Src : Zero;
+
+    Result.Low = _VExtr(OpSize::i128Bit, OpSize::i8Bit, Src1.Low, Src2.Low, ExtrShift);
+    if (!Is128Bit) {
+      Result.High = _VExtr(OpSize::i128Bit, OpSize::i8Bit, Src1.High, Src2.High, ExtrShift);
+    }
+  }
+
+  if (Is128Bit) {
+    Result.High = LoadZeroVector(OpSize::i128Bit);
+  }
+
+  AVX128_StoreResult_WithOpSize(Op, Op->Dest, Result);
+}
+
+void OpDispatchBuilder::AVX128_VPSRLDQ(OpcodeArgs) {
+  AVX128_ShiftDoubleImm(Op, true /* Right */);
+}
+
+void OpDispatchBuilder::AVX128_VPSLLDQ(OpcodeArgs) {
+  AVX128_ShiftDoubleImm(Op, false /* Right */);
```

Comment:
```suggestion
  AVX128_ShiftDoubleImm(Op, false /* Left */);
```

Reply:
I think that might be more confusing. it is false that this is Right. it is true that this is Left.

Reply:
Conversely I find that extremely confusing. The programmatic part should absolutely not be inverting a comment lol, and it genuinely comes across as a copy-paste typo to me. This can be more explicit by just making the comments "Right shift" and "Left shift".

Reply:
Easy fix for this is to do a simple enum
```cpp
enum class ShiftDirection {
  RIGHT, LEFT,
};

void AVX128_ShiftDoubleImm(OpcodeArgs, ShiftDirection Dir);

void OpDispatchBuilder::AVX128_VPSRLDQ(OpcodeArgs) {
  AVX128_ShiftDoubleImm(Op, ShiftDirection::RIGHT);
}

void OpDispatchBuilder::AVX128_VPSLLDQ(OpcodeArgs) {
  AVX128_ShiftDoubleImm(Op, ShiftDirection::LEFT);
}
```

Reply:
We have a winner!